Library/Homebrew/test/utils_spec.rb

```diff
@@ -0,0 +1,340 @@
+require "utils"
+require "tempfile"
+require "utils/shell"
+
+RSpec::Matchers.alias_matcher :have_failed, :be_failed
+
+def esc(code)
+  /(\e\[\d+m)*\e\[#{code}m/
+end
+
+module Utils
+  describe self do
+    let(:dir) { Pathname.new(Dir.mktmpdir) }
+
+    specify "#ofail" do
+      begin
+        shutup do
+          ofail "foo"
+        end
+
+        expect(Homebrew).to have_failed
+      ensure
+        Homebrew.failed = false
+      end
+    end
+
+    specify "#odie" do
+      expect(self).to receive(:exit).and_return(1)
+      shutup do
+        odie "foo"
+      end
+    end
+
+    describe "#pretty_installed" do
+      context "with HOMEBREW_NO_EMOJI unset and when $stdout is a TTY" do
+        it "returns a string with a colored checkmark" do
+          allow($stdout).to receive(:tty?).and_return(true)
+          ENV.delete("HOMEBREW_NO_EMOJI")
+          expect(pretty_installed("foo")).to match(/#{esc 1}foo #{esc 32}✔#{esc 0}/)
+        end
+      end
+
+      context "with HOMEBREW_NO_EMOJI set and when $stdout is a TTY" do
+        it "returns a string with colored info" do
+          allow($stdout).to receive(:tty?).and_return(true)
+          ENV["HOMEBREW_NO_EMOJI"] = "1"
+          expect(pretty_installed("foo")).to match(/#{esc 1}foo \(installed\)#{esc 0}/)
+        end
+      end
+
+      context "when $stdout is not a TTY" do
+        it "returns plain text" do
+          allow($stdout).to receive(:tty?).and_return(false)
+          expect(pretty_installed("foo")).to eq("foo")
+        end
+      end
+    end
+
+    describe "#pretty_uninstalled" do
+      context "with HOMEBREW_NO_EMOJI unset and when $stdout is a TTY" do
+        it "returns a string with a colored checkmark" do
+          allow($stdout).to receive(:tty?).and_return(true)
+          ENV.delete("HOMEBREW_NO_EMOJI")
+          expect(pretty_uninstalled("foo")).to match(/#{esc 1}foo #{esc 31}✘#{esc 0}/)
+        end
+      end
+
+      context "with HOMEBREW_NO_EMOJI set and when $stdout is a TTY" do
+        it "returns a string with colored info" do
+          allow($stdout).to receive(:tty?).and_return(true)
+          ENV["HOMEBREW_NO_EMOJI"] = "1"
+          expect(pretty_uninstalled("foo")).to match(/#{esc 1}foo \(uninstalled\)#{esc 0}/)
+        end
+      end
+
+      context "when $stdout is not a TTY" do
+        it "returns plain text" do
+          allow($stdout).to receive(:tty?).and_return(false)
+          expect(pretty_uninstalled("foo")).to eq("foo")
+        end
+      end
+    end
+
+    describe "#interactive_shell" do
+      let(:shell) { dir/"myshell" }
+
+      after(:each) do
+        dir.rmtree
+      end
```

Comment:
I think this should be done at the same scope `dir` is declared on. You can prevent accidentally initialising `dir` in the `after` block if it hasn't been used in the test like this:

```ruby
let(:dir) { @dir = Dir.mktmpdir }
after { @dir.unlink if @dir }
```